README
rotter-web-service Readme
=========================
This article describes the rotter-web-service API and uses the following
conventions: Code is presented in code blocks, mandatory input parameters are
enclosed in angle brackets "<" and ">" and optional input parameters are
enclosed in square brackets "[" and "]".

All input parameters can be either GET or POST variables. '''Unless otherwise
stated, times are UTC.'''

listservices.php:
-----------------
; Service Description: Lists the services that can be queried through the various AudioFile Web Services

; Input Parameters: none required
; [callback]: JavaScript function name

; Returned Output: A JSON list of services in the format:
```JSON
 {
 	"services":[
 		{
                         "title":"RTÉ Radio 1 FM",
                         "id":"radio1"
 		},
 		{
 			"title":"RTÉ 2fm",
 			"id":"2fm"
 		},
 		{
 			"title":"RTÉ lyric fm",
 			"id":"lyricfm"
 		},
 		...
 	]
 }
```

listfiles.php:
--------------
; Service Description: Lists all of the files for a given service on a given date.

; Input Parameters
; <service>: one of the ids returned by listservices.php
; <date>: in format "YYYY-MM-DD"

; Returned Output: A JSON list of files in the format:
```JSON
 {
 	"files":[
 		{
 			"title":"00:00:00",
 			"file":"2011-01-18-00-00-00-00.mp3",
 			"size":28802317
 		},
 		{
 			"title":"01:00:00",
 			"file":"2011-01-18-01-00-00-02.mp3",
 			"size":28803137
 		}
 	]
 }
```

'''Note:''' The title field in the returned JSON is adjusted for daylight saving time.

download.php:
-------------
; Service Description: Produce a single file for download from the specified service between the start and end times specified.

; Input Parameters
; <service>: one of the ids returned by listservices.php
; <start>: in format "YYYY-MM-DD-HH-mm-ss-hh", hh represents the hundredth of a second
; <end>: in format "YYYY-MM-DD-HH-mm-ss-hh"
; [format]: "mp2" or "mp3" - mp3 by default
; [file_title]: the title returned in the 'Content-Disposition: attachment; filename=' HTTP header (exclude the file suffix (i.e. don't include the '.mp2' - this is set automatically). If not provided, this defaults to 'output'

'''Note:''' The maximum difference between start and end times has been capped and tested at 24 hours.

'''Note:''' The start and end times can straddle day boundaries.

; Returned Output: A HTTP 200 response with Content-Type: audio/mpeg followed by the mpeg audio data.
;: If there is any problem, a HTTP 404 error will be generated.


CHANGELOG
2011-11-15
----------
Created dst-test.php to diagnose DST problems

The file higlights if the timezone setting in php.ini is set correctly.
Also added a file_name input parameter to the download API for arbitrary download file titles.
Added a test service into the hacked listservices API.

2011-08-12
----------
Reduced logging level in download API.
Hacked listservices API in the absence of time to do it better.

2011-07-22
----------
Added a UTC correction into the times that are returned by listfiles.php

2011-07-08
----------
Added readfile_chunked function.
Made other minor changes, including ones to common-config.php.
Fixed JSON formatting errors in listfiles.php and listservices.php.
Added JSONP callback parameter to listfiles.php.


dst-test.php
<?php

$files = array(
'2011-10-29-20-00-00-00.mp3',
'2011-10-29-21-00-00-00.mp3',
'2011-10-29-22-00-00-00.mp3',
'2011-10-29-23-00-00-00.mp3',
'2011-10-30-00-00-00-02.mp3',
'2011-10-30-01-00-00-00.mp3',
'2011-10-30-02-00-00-01.mp3',
'2011-10-30-03-00-00-01.mp3',
'2011-10-30-04-00-00-00.mp3',
'2011-10-30-05-00-00-00.mp3',
'2011-10-30-06-00-00-00.mp3',
'2011-10-30-07-00-00-00.mp3',
'2011-10-30-08-00-00-01.mp3',
'2011-10-30-09-00-00-00.mp3',
'2011-10-30-10-00-00-02.mp3',
'2011-10-30-11-00-00-01.mp3',
'2011-10-30-12-00-00-00.mp3'
);

?>
<html>
	<head>
	<title>Test</title>
	<style type="text/css">
table {
	border: solid navy 1px
}
thead {
	font-weight: bold
}
td {
	border: solid silver 1px;
	padding: 0.1em
}
	</style>
	</head>
<body>
<table>
	<thead>
		<td>File:</td>
		<td>Unix Time:</td>
		<td>tm_isdst</td>
		<td>date('H:i:s e')</td>
	</thead>
<?php

foreach ( $files as $file ) {
	list($year, $month, $day, $hour, $minute, $second, $hundredth) = explode('-', $file);
	$hundredth = substr($hundredth, 0, 2);

	$unix_time = mktime( $hour, $minute, $second, $month, $day, $year, 0 );
	$is_dst = localtime( $unix_time, true );
	$is_dst = $is_dst['tm_isdst'];
	$date = date('Y-m-d H:i:s e', $unix_time);
	echo "	<tr>\n";
	echo "		<td>$file</td>
		<td>$unix_time</td>
		<td>$is_dst</td>
		<td>$date</td>\n";
	echo "	</tr>\n";
}

?>
</table>
</body>
</html>


index.php
<?php
/*
 * Created on Oct 20, 2010
 *
 * To change the template for this generated file go to
 * Window - Preferences - PHPeclipse - PHP - Code Templates
 */

// Mpgedit binary
$mpgedit = '/usr/local/bin/mpgedit_nocurses';
//$mpgedit = '/bin/echo';

// The base directory for rotter recordings
$rotter_base_dir = '/var/rotter/';

// The suffix on the recording filenames
$recording_suffix = '.mp2';

// Arguments we want to get and any defaults we want to set
// Service
$service = isset($_GET['service']) ? $_GET['service'] : ( isset($_POST['service']) ? $_POST['service'] : 'radio1' );

// All dates and times are in UTC
date_default_timezone_set('UTC');

// Start Date and Time (Date = YYYY-MM-DD, Time = HH-mm-ss-hh - UTC)
$request_start = isset($_GET['start']) ? $_GET['start'] : ( isset($_POST['start']) ? $_POST['start'] : gmdate("Y-m-d-H-i-s-00", time()-300) );
//$request_start = '2010-10-22-09-00-00-00';
preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))$/', $request_start, $matches);
list(, $start_date, $start_yr, $start_mo, $start_dy, $start_time, $start_hr, $start_mi, $start_se, $start_hs) = $matches;

// End Date and Time
$request_end = isset($_GET['end']) ? $_GET['end'] : ( isset($_POST['end']) ? $_POST['end'] : gmdate("Y-m-d-H-i-s-00") );
//$request_end = '2010-10-22-10-00-00-01';
preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))$/', $request_end, $matches);
list(, $end_date, $end_yr, $end_mo, $end_dy, $end_time, $end_hr, $end_mi, $end_se, $end_hs) = $matches; 

// Before doing anything else, is the request for anything over 24 hours?
if ( get_time_diff($request_start, $request_end) > 86400 ) {
	die('You cannot request audio longer than 24 hours long - That would be ridiculous.');
}

// Here's the science!!
/*
 * a - the start date_time of the first useful file 
 * b - the requested start date_time
 * ( -e(b-a)- )
 * 
 * y - the start date_time of the last useful file
 * z - the requested end date_time
 * ( -e-(z-y) )
 */

function dir_listing_filter($file) {
	global $recording_suffix;
	
	$pmatch = '/^\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}'. $recording_suffix .'/';
	if ( preg_match($pmatch, $file) != 1 ) {
		return false;
	} else {
		return true;
	}
}

/* For an end date_time that corresponds to a filename,
 * should the editspec for mpgedit be: -e - ?
 * 
 * Also, I need to check if there is a way to determine the length of a file quickly so that
 * missing patches can be detected and replaced with silence.
 */
function get_filename_for_date_time($date_time, $is_end = false) {
	global $rotter_base_dir, $service, $recording_suffix;

	// Get the date out of date_time
	preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))$/', $date_time, $matches);
	list(, $date, $yr, $mo, $dy, $time, $hr, $mi, $se, $hs) = $matches;

	// Get date_time - 1 day
	$day_before_date_time = strftime("%Y-%m-%d", gmmktime(0, 0, 0, $mo, $dy-1, $yr));

	// List the directory for date_time
	$dir_listing = scandir($rotter_base_dir . $service . '/' . $date);
	
	// Add yesterday's directory listing too
	$dir_listing = array_merge($dir_listing, scandir($rotter_base_dir . $service . '/' . $day_before_date_time));
	
	// Clear out any unwanted files ('.idx', '.', '..', etc.)
	$dir_listing = array_filter($dir_listing, "dir_listing_filter");

	// Insert our date_time as a fake file into the directory listing
	$fake_file = str_replace(array(' ', ':', '.'), '-', $date_time) . $recording_suffix;
	$dir_listing[] = $fake_file;

	// Reverse sort the array so that index + 1 will be the file we want
	rsort($dir_listing);
	
	// The most likely candidate
	$file = $dir_listing[array_search($fake_file, $dir_listing)+1];
	
	// If is_end is set and the timestamp of the file is the same as that of the fake file, we want the next earliest file
	if ( $file == $date_time.$recording_suffix && $is_end === true ) {
		$file = $dir_listing[array_search($fake_file, $dir_listing)+2];
	}
	
	return $file;
}

function get_intermediate_files($file_a, $file_y) {
	global $rotter_base_dir, $service, $recording_suffix;

	preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))' . $recording_suffix . '$/', $file_a, $matches);
	list(, $start_date, $start_yr, $start_mo, $start_dy, $start_time, $start_hr, $start_mi, $start_se, $start_hs) = $matches;
	
	preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))' . $recording_suffix . '$/', $file_y, $matches);
	list(, $end_date, $end_yr, $end_mo, $end_dy, $end_time, $end_hr, $end_mi, $end_se, $end_hs) = $matches; 
	
	// Scan the relavant directory / directories
	$dir_listing = scandir("$rotter_base_dir/$service/$start_date");
	if ( $start_date != $end_date ) $dir_listing = array_merge($dir_listing, scandir("$rotter_base_dir/$service/$end_date"));
	
	// Clear out any unwanted files ('.idx', '.', '..', etc.)
	$dir_listing = array_filter($dir_listing, "dir_listing_filter");

	rsort($dir_listing);
	$index = count($dir_listing)-1;
	while ( $dir_listing[$index] != $file_a ) {
		array_pop($dir_listing);
		$index--;
	}
	array_pop($dir_listing);

	sort($dir_listing);
	$index = count($dir_listing)-1;
	while ( $dir_listing[$index] != $file_y ) {
		array_pop($dir_listing);
		$index--;
	}
	array_pop($dir_listing);
	
	return $dir_listing;
}

function get_time_diff($date_time1, $date_time2) {
	// Accepts date_times in the form:
	// YYYY-MM-DD-HH-mm-ss-hh or
	// YYYY-MM-DD HH:mm:ss.hh
	
	preg_match('/^(\d{4})-(\d{2})-(\d{2})[ -](\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2})$/', $date_time1, $matches);
	list(, $yr, $mo, $dy, $hr, $mi, $se, $hs) = $matches;
	$days1 = gregoriantojd($mo, $dy, $yr);
	$time1 = floatval(intval($hr)*3600 + intval($mi)*60 + intval($se) . '.' . $hs);
	
	preg_match('/^(\d{4})-(\d{2})-(\d{2})[ -](\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2})$/', $date_time2, $matches);
	list(, $yr, $mo, $dy, $hr, $mi, $se, $hs) = $matches;
	$days2 = gregoriantojd($mo, $dy, $yr);
	$time2 = floatval(intval($hr)*3600 + intval($mi)*60 + intval($se) . '.' . $hs);
	
	$days_diff = $days2 - $days1;
	$time_diff = $time2 - $time1;
	
	return round($days_diff*3600*24 + $time_diff, 2);
}

// Get the file that contains the start
$file_a = get_filename_for_date_time($request_start);

// How far in is the start we want
$offset_start = get_time_diff(str_replace($recording_suffix, '', $file_a), $request_start);

// Get the file that contains the end
$file_y = get_filename_for_date_time($request_end, true);

// How far in is the end we want
$offset_end = get_time_diff(str_replace($recording_suffix, '', $file_y), $request_end);

// Get any "in between" files
$files = $file_a != $file_y ? get_intermediate_files($file_a, $file_y) : array();

// Prepare a temp file
$temp_file = tempnam('/tmp/', 'mpgedit_op_');
// Check for faillure
if ($temp_file === FALSE) die('Failed to create temporary file!');
// mpgedit doesn't like overwriting files, so...
unlink($temp_file);

// Build the command line we'll use
$cmd = "$mpgedit -o $temp_file -e $offset_start-";
if ( $file_a != $file_y ) {
	// Start and end are in different files
	$cmd .= " -f $rotter_base_dir$service/$start_date/$file_a";
	foreach ( $files as $file ) {
		// For every intermediary file, we can just append the whole thing
		preg_match('/^\d{4}-\d{2}-\d{2}/', $file, $matches);
		$file_date = $matches[0];
		$cmd .= " -e - -f $rotter_base_dir$service/$file_date/$file";
	}
	// Because the end is in a different file to the beginning, we must restart it's editspec
	$cmd .= " -e -";
}
$cmd .= "$offset_end -f $rotter_base_dir$service/$end_date/$file_y";

// Run the mpgedit application (timespecs can be in [mm:]ss[.hh] format)
exec($cmd, $mpgedit_op, $ret_val);

if ($ret_val == 0) {
	//header('Expires: ' . gmdate('D, d M Y H:i:s e'));
	header('Content-Type: audio/mpeg');
	header('Content-Disposition: attachment; filename=output' . $recording_suffix);
	header('Content-Length: ' . filesize($temp_file));
	ini_set('memory_limit', '-1');
	set_time_limit(0);
	readfile($temp_file);
	//echo $temp_file;
	//var_dump($ret_val, $mpgedit_op);
	//echo 'Content-Length: '.filesize($temp_file)."<br />\n";
} else {
	var_dump($mpgedit_op);
	echo "<br />Dumpfh! - You landed here!";
exit;
}

// Remove the file 
if ( file_exists($temp_file) ) unlink($temp_file);

?>


listfiles.php
<?php

// Include the common configuration parameters
require_once('common-config.php');

// Parse the service from the GET or POST variables
$service = isset($_GET['service']) ? $_GET['service'] : ( isset($_POST['service']) ? $_POST['service'] : 'radio1' );

// Parse the date from the GET or POST variables
$date = isset($_GET['date']) ? $_GET['date'] : ( isset($_POST['date']) ? $_POST['date'] : date('Y-m-d') );

// Is a callback function referenced?
$callback = isset($_GET['callback']) ? $_GET['callback'] : ( isset($_POST['callback']) ? $_POST['callback'] : false );

// Get the directory listing and safely exit if there's a problem
if ( ($dir_listing = scandir($rotter_base_dir . $service . '/' . $date)) === FALSE) die('{
	"files":[
	]
}');

// Loop through and only keep the relevant audio files
$i = 0;
while ( $i < count($dir_listing) ) {
	if ( preg_match('/^\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}' . $recording_suffix . '$/', $dir_listing[$i]) != 1 ) {
		array_splice($dir_listing, $i, 1);
	} else {
		$i++;
	}
}

// Is there a callback?
if ( $callback !== false ) {
	echo $callback, '(';
}

// Begin Output
echo '{
	"files":[
';

// Per file output
for ($i = 0; $i < count($dir_listing); $i++) {
	preg_match('/^(\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2})-\d{2}' . $recording_suffix . '$/', $dir_listing[$i], $matches);
	list($year, $month, $day, $hour, $minute, $second) = explode('-', $matches[1]);
	$time = mktime( $hour, $minute, $second, $month, $day, $year, 0 );
	$title = date("H:i:s", $time);
	echo '		{
			"title":"' . $title . '",
			"file":"' . $dir_listing[$i] . '",
			"size":' . filesize($rotter_base_dir . $service . '/' . $date . '/' . $dir_listing[$i]) . '
		}';
	if ($i != count($dir_listing)-1) {
		echo ',';
	}
	echo "\n";
}

// End output
echo '	]
}';

// Is there a callback?
if ( $callback !== false ) {
        echo ');';
}


?>


listservices.php
<?php

// Is a callback function referenced?
$callback = isset($_GET['callback']) ? $_GET['callback'] : ( isset($_POST['callback']) ? $_POST['callback'] : false );

// Is there a callback?
if ( $callback !== false ) {
	echo $callback, '(';
}

?>{
	"services":[
		{
			"title":"RTÉ Radio 1 FM",
			"id":"radio1"
		},
		{
			"title":"RTÉ 2fm",
			"id":"2fm"
		},
		{
			"title":"RTÉ lyric fm",
			"id":"lyricfm"
		},
		{
			"title":"RTÉ Raidió na Gaeltachta",
			"id":"rnag"
		},
		{
			"title":"RTÉ Gold",
			"id":"gold"
		},
		{
			"title":"RTÉ 2XM",
			"id":"2xm"
		},
		{
			"title":"RTÉ Choice",
			"id":"choice"
		},
		{
			"title":"RTÉ Junior/Chill",
			"id":"junior-chill"
		},
		{
			"title":"RTÉ Pulse",
			"id":"pulse"
		},
		{
			"title":"RTÉ Radio 1 extra",
			"id":"radio1extra"
		},
		{
			"title":"RTÉ Radio 1 LW",
			"id":"radio1lw"
		},
		{
			"title":"Newstalk",
			"id":"newstalk"
		},
		{
			"title":"Today fm",
			"id":"todayfm"
		},
		{
			"title":"4fm",
			"id":"4fm"
		},
		{
			"title":"Test",
			"id":"radio1test"
		}
	]
}<?php

// Is there a callback?
if ( $callback !== false ) {
	echo ');';
}

?>


download.php
<?php
/*
 * Created on Oct 20, 2010
 *
 * To change the template for this generated file go to
 * Window - Preferences - PHPeclipse - PHP - Code Templates
 */

// All dates and times are in UTC
date_default_timezone_set('UTC');

// Log levels
// 0 - nothing
// 1 - fatal errors
// 2 - one line record
// 3 - operations
// 4 - operations incl. repeated ones
// 5 - lots of info
$log_level = 3;

// Logging function
function log_message($msg_level, $msg, $bare_log = false) {
	global $log_level;

	// Should this go any further
	if ($log_level < $msg_level) return true;
        // open file
        $fd = fopen(dirname($_SERVER['SCRIPT_FILENAME']) . '/download.log', 'a');
	if (!$fd) return false;
        // append date/time to message
        $str = ($bare_log === false ? '[' . date('Y/m/d H:i:s', mktime()) . '] ' . $_SERVER['REMOTE_ADDR'] . ' [' . $msg_level . '] ' : '') . $msg;
        // write string
        if (fwrite($fd, $str . "\n") === false) return false;
        // close file
        fclose($fd);
	return true;
}

// Include Common configuration parameters
require_once('common-config.php');

// Mpgedit binary
$mpgedit = '/usr/local/bin/mpgedit_nocurses';
//$mpgedit = '/bin/echo';

// Arguments we want to get and any defaults we want to set
// Service
$service = isset($_GET['service']) ? $_GET['service'] : ( isset($_POST['service']) ? $_POST['service'] : 'radio1' );
log_message(4, 'Service: ' . $service);

// Start Date and Time (Date = YYYY-MM-DD, Time = HH-mm-ss-hh - UTC)
$request_start = isset($_GET['start']) ? $_GET['start'] : ( isset($_POST['start']) ? $_POST['start'] : gmdate("Y-m-d-H-i-s-00", time()-300) );
//$request_start = '2010-10-22-09-00-00-00';
preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))$/', $request_start, $matches);
list(, $start_date, $start_yr, $start_mo, $start_dy, $start_time, $start_hr, $start_mi, $start_se, $start_hs) = $matches;
log_message(4, 'Start date: ' . $start_date . ', time: ' . $start_time);

// End Date and Time
$request_end = isset($_GET['end']) ? $_GET['end'] : ( isset($_POST['end']) ? $_POST['end'] : gmdate("Y-m-d-H-i-s-00") );
//$request_end = '2010-10-22-10-00-00-01';
preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))$/', $request_end, $matches);
list(, $end_date, $end_yr, $end_mo, $end_dy, $end_time, $end_hr, $end_mi, $end_se, $end_hs) = $matches; 
log_message(4, 'End date: ' . $end_date . ', time: ' . $end_time);

// Title for the downloaded clip
$file_title = isset($_GET['file_title']) ? $_GET['file_title'] : ( isset($_POST['file_title']) ? $_POST['file_title'] : 'output' );

// Before doing anything else, is the request for anything over 24 hours?
if ( get_time_diff($request_start, $request_end) > 86400 ) {
	log_message(1, 'Request for audio file larger than 24 hours - die()ing');
	die('You cannot request audio longer than 24 hours long - That would be ridiculous.');
}

/**
 * dir_listing_filter - returns true only if file is an audio file that we care about
 */
function dir_listing_filter($file) {
	global $recording_suffix;
	log_message(5, 'dir_listing_filter($file = \'' . $file . '\')');

	$pmatch = '/^\d{4}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}'. $recording_suffix .'/';
	if ( preg_match($pmatch, $file) != 1 ) {
		return false;
	} else {
		return true;
	}
}

/* For an end date_time that corresponds to a filename,
 * should the editspec for mpgedit be: -e - ?
 * 
 * Also, I need to check if there is a way to determine the length of a file quickly so that
 * missing patches can be detected and replaced with silence.
 */
function get_filename_for_date_time($date_time, $is_end = false) {
	global $rotter_base_dir, $service, $recording_suffix;
	log_message(4, 'get_filename_for_date_time($date_time = \'' . $date_time . '\', $is_end = \'' . $is_end . '\')');

	// Get the date out of date_time
	preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))$/', $date_time, $matches);
	list(, $date, $yr, $mo, $dy, $time, $hr, $mi, $se, $hs) = $matches;

	// List the directory for date_time
	$dir_listing = scandir($rotter_base_dir . $service . '/' . $date);

	// Get date_time - 1 day
	$day_before_date_time = strftime("%Y-%m-%d", gmmktime(0, 0, 0, $mo, $dy-1, $yr));

	// Get yesterday's directory listing if it exists
	$yday_dir_listing = scandir($rotter_base_dir . $service . '/' . $day_before_date_time);

	// Add yesterday's directory listing too if it exists
	if ( $yday_dir_listing !== false ) $dir_listing = array_merge($dir_listing, $yday_dir_listing);

	// Clear out any unwanted files ('.idx', '.', '..', etc.)
	$dir_listing = array_filter($dir_listing, "dir_listing_filter");

	// Insert our date_time as a fake file into the directory listing
	$fake_file = str_replace(array(' ', ':', '.'), '-', $date_time) . $recording_suffix;
	$dir_listing[] = $fake_file;

	// Reverse sort the array so that index + 1 will be the file we want
	rsort($dir_listing);

	// The most likely candidate
	$file = $dir_listing[array_search($fake_file, $dir_listing)+1];

	// If is_end is set and the timestamp of the file is the same as that of the fake file, we want the next earliest file
	if ( $file == $date_time.$recording_suffix && $is_end === true ) {
		$file = $dir_listing[array_search($fake_file, $dir_listing)+2];
	}

	return $file;
}

function get_intermediate_files($file_a, $file_y) {
	global $rotter_base_dir, $service, $recording_suffix;
	log_message(4, 'get_intermediate_files($file_a = \'' . $file_a . '\', $file_b = \'' . $file_y . '\')');

	preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))' . $recording_suffix . '$/', $file_a, $matches);
	list(, $start_date, $start_yr, $start_mo, $start_dy, $start_time, $start_hr, $start_mi, $start_se, $start_hs) = $matches;
	
	preg_match('/^((\d{4})-(\d{2})-(\d{2}))[ -]((\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2}))' . $recording_suffix . '$/', $file_y, $matches);
	list(, $end_date, $end_yr, $end_mo, $end_dy, $end_time, $end_hr, $end_mi, $end_se, $end_hs) = $matches; 
	
	// Scan the relavant directory / directories
	$dir_listing = scandir("$rotter_base_dir/$service/$start_date");
	if ( $start_date != $end_date ) $dir_listing = array_merge($dir_listing, scandir("$rotter_base_dir/$service/$end_date"));
	
	// Clear out any unwanted files ('.idx', '.', '..', etc.)
	$dir_listing = array_filter($dir_listing, "dir_listing_filter");

	rsort($dir_listing);
	$index = count($dir_listing)-1;
	while ( $dir_listing[$index] != $file_a ) {
		array_pop($dir_listing);
		$index--;
	}
	array_pop($dir_listing);

	sort($dir_listing);
	$index = count($dir_listing)-1;
	while ( $dir_listing[$index] != $file_y ) {
		array_pop($dir_listing);
		$index--;
	}
	array_pop($dir_listing);
	
	return $dir_listing;
}

function get_time_diff($date_time1, $date_time2) {
	// Accepts date_times in the form:
	// YYYY-MM-DD-HH-mm-ss-hh or
	// YYYY-MM-DD HH:mm:ss.hh
	log_message(4, 'get_time_diff($date_time1 = \'' . $date_time1 . '\', $date_time2 = \'' . $date_time2 . '\')');

	preg_match('/^(\d{4})-(\d{2})-(\d{2})[ -](\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2})$/', $date_time1, $matches);
	list(, $yr, $mo, $dy, $hr, $mi, $se, $hs) = $matches;
	$days1 = gregoriantojd($mo, $dy, $yr);
	$time1 = floatval(intval($hr)*3600 + intval($mi)*60 + intval($se) . '.' . $hs);

	preg_match('/^(\d{4})-(\d{2})-(\d{2})[ -](\d{2})[:-](\d{2})[:-](\d{2})[.-](\d{2})$/', $date_time2, $matches);
	list(, $yr, $mo, $dy, $hr, $mi, $se, $hs) = $matches;
	$days2 = gregoriantojd($mo, $dy, $yr);
	$time2 = floatval(intval($hr)*3600 + intval($mi)*60 + intval($se) . '.' . $hs);

	$days_diff = $days2 - $days1;
	$time_diff = $time2 - $time1;

	return round($days_diff*3600*24 + $time_diff, 2);
}

// Here's the science!!
/*
 * a - the start date_time of the first useful file 
 * b - the requested start date_time
 * ( -e(b-a)- )
 * 
 * y - the start date_time of the last useful file
 * z - the requested end date_time
 * ( -e-(z-y) )
 */

// Get the file that contains the start
$file_a = get_filename_for_date_time($request_start);

// How far in is the start we want
$offset_start = get_time_diff(str_replace($recording_suffix, '', $file_a), $request_start);

// Get the file that contains the end
$file_y = get_filename_for_date_time($request_end, true);

// How far in is the end we want
$offset_end = get_time_diff(str_replace($recording_suffix, '', $file_y), $request_end);

// Get any "in between" files
$files = $file_a != $file_y ? get_intermediate_files($file_a, $file_y) : array();

// Prepare a temp file
$temp_file = tempnam('/tmp/', 'mpgedit_op_');
// Check for faillure
if ($temp_file === FALSE) {
	log_message(1, 'Failed to create temporary file');
	die('Failed to create temporary file!');
}
// mpgedit doesn't like overwriting files, so...
unlink($temp_file);
log_message(3, 'File: ' . $temp_file . ' allocated for audio data');

// This needs to run when the script is finnishing
function tidy_up_func($temp_file) {
	// This message never gets logged
	log_message(3, 'Deleting file: ' . $temp_file);
	// Remove the file
	if ( file_exists($temp_file) ) unlink($temp_file);
}
register_shutdown_function('tidy_up_func', $temp_file);

// Build the command line we'll use
$cmd = "$mpgedit -o $temp_file -e $offset_start-";
log_message(4, 'Command line: ' . $cmd);

if ( $file_a != $file_y ) {
	// Start and end are in different files
	$cmd .= " -f $rotter_base_dir$service/$start_date/$file_a";
	log_message(4, "\t:  -f $rotter_base_dir$service/$start_date/$file_a", true);
	foreach ( $files as $file ) {
		// For every intermediary file, we can just append the whole thing
		preg_match('/^\d{4}-\d{2}-\d{2}/', $file, $matches);
		$file_date = $matches[0];
		$cmd .= " -e - -f $rotter_base_dir$service/$file_date/$file";
		log_message(4, "\t:  -e - -f $rotter_base_dir$service/$file_date/$file", true);
	}
	// Because the end is in a different file to the beginning, we must restart it's editspec
	$cmd .= " -e -";
	log_message(4, "\t:  -e -", true);
}
$cmd .= "$offset_end -f $rotter_base_dir$service/$end_date/$file_y";
log_message(4, "\t: $offset_end -f $rotter_base_dir$service/$end_date/$file_y", true);

// Not sure if these are actually needed - seemed to be working ok without
ini_set('memory_limit', '-1');
set_time_limit(0);

// Run the mpgedit application (timespecs can be in [mm:]ss[.hh] format)
log_message(2, 'Executing: ' . $cmd);
exec($cmd, $mpgedit_op, $ret_val);

function readfile_chunked($filename, $retbytes=true) {
	$chunksize = 1*(1024*1024); // how many bytes per chunk
	$buffer = '';
	$cnt =0;
	// $handle = fopen($filename, 'rb');
	$handle = fopen($filename, 'rb');
	if ($handle === false) {
		return false;
	}
	while (!feof($handle)) {
		$buffer = fread($handle, $chunksize);
		echo $buffer;
		ob_flush();
		flush();
		if ($retbytes) {
			$cnt += strlen($buffer);
		}
	}
	$status = fclose($handle);
	if ($retbytes && $status) {
		return $cnt; // return num. bytes delivered like readfile() does.
	}
	return $status;

}

if ($ret_val == 0) {
	log_message(5, "$mpgedit output:\n" . $mpgedit_op);
	//header('Expires: ' . gmdate('D, d M Y H:i:s e'));
	header('Content-Type: audio/mpeg');
	header('Content-Disposition: attachment; filename=' . $file_title . $recording_suffix);
	header('Content-Length: ' . filesize($temp_file));
	log_message(3, 'About to readfile_chunked()');
	readfile_chunked($temp_file);
	exit;
} else {
	log_message(1, $mpgedit . ' error: return code: ' . $ret_val);
	log_message(4, $mpgedit_op);
	echo 'There was a problem creating the audio file.';
	exit;
}

?>
